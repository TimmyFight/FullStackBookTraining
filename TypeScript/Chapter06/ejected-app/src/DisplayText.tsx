import React, { useState, FC } from "react";
import UserTodos from "./UserTodos"

interface DisplayTextProps {
  getUserFullName: (username: string) => Promise<string>;
}

const DisplayText: FC<DisplayTextProps> = ({ getUserFullName }) => {
  const [txt, setTxt] = useState("");
  const [msg, setMsg] = useState("");

  const [todoControl, setTodoControl] = useState<ReturnType<typeof UserTodos>>();

  const onChangeTxt = (e: React.ChangeEvent<HTMLInputElement>) => {
    setTxt(e.target.value);
  };

  const onClickShowMsg = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault();
    setMsg(`Welcome on the React Testing training, ${await getUserFullName(txt)}!`);
    setTodoControl(<UserTodos username={txt} />);
  };

  return (
    <form>
      <div>
        <label>What is your name?</label>
      </div>
      <div>
        <input data-testid="user-input" value={txt} onChange={onChangeTxt} />
      </div>
      <div>
        <button data-testid="input-submit" onClick={onClickShowMsg}>Show the message</button>
      </div>
      <div>
        <label data-testid="final-msg">{msg}</label>
      </div>
      <div>
        Elemnt for testing purposes only
      </div>
      <ul style={{ marginTop: '1rem', listStyleType: 'none' }}>
        { todoControl }
      </ul>
    </form>
  );
}

export default DisplayText;
