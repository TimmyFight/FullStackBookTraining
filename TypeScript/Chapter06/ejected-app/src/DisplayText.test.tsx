import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import DisplayText from "./DisplayText";
import "@testing-library/jest-dom/extend-expect";

describe("Testing DisplayText component", () => {
  it("Component is rendered without any issues", () => {
    const {baseElement} = render(<DisplayText />);
    console.log(baseElement.innerHTML);
    expect(baseElement).toBeInTheDocument();
  });

  it("Compatible with snapshot", () => {
    const {baseElement} = render(<DisplayText />);
    expect(baseElement).toMatchSnapshot();
  });

  it("Given value is available for component", () => {
    const testUser = "Test User";
    render(<DisplayText />);
    const input = screen.getByTestId("user-input");
    fireEvent.change(input, { target: { value: testUser }});
    expect(input).toBeInTheDocument();
    expect(input).toHaveValue(testUser);
  });

  it("Welcome message has been print on the screen", () => {
    const testUser = "Test User";
    const message = `Welcome on the React Testing training, ${testUser}!`;
    
    render (<DisplayText />);

    const input = screen.getByTestId("user-input");
    const label = screen.getByTestId("final-msg");

    fireEvent.change(input, { target: { value: testUser } });

    const btn = screen.getByTestId("input-submit");
    fireEvent.click(btn);

    expect(label).toBeInTheDocument();
    expect(label.innerHTML).toBe(message);
  });
});
