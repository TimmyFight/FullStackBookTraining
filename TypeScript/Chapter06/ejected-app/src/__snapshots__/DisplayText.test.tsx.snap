// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing DisplayText component Compatible with snapshot 1`] = `
<body>
  <div>
    <form>
      <div>
        <label>
          What is your name?
        </label>
      </div>
      <div>
        <input
          data-testid="user-input"
          value=""
        />
      </div>
      <div>
        <button
          data-testid="input-submit"
        >
          Show the message
        </button>
      </div>
      <div>
        <label
          data-testid="final-msg"
        />
      </div>
      <div>
        Elemnt for testing purposes only
      </div>
    </form>
  </div>
</body>
`;
